### DISCUSSION TITLE
Snapshots which can be saved and loaded later.

### DISCUSSION THREAD

**Discussion by rhinoid** [dirkwhoffmann/vAmiga#761] [Ideas]
At the moment it looks like snapshots are ephemeral, and they go away the moment the machine is turned off.
So the use case right now is a way to skip back in a game and have infinite tries.

The use case I'm thinking of is to have snapshots of the various interesting points in my game(s) so people can jump in without first playing everything. So optimally I'd play on the Mac and curate some snapshots, and later would be able to load them up with the webplayer if someone clicks a hyperlink with an image or a text.

**Comment by mithrendal** [comment]
nice idea... you would give the visitor of your page an `instant on` of your game without the usual disk change procedure when a game has multiple disks to load

but there are these obstacles....

1. snapshot format might not be compatible between the mac version and the webassembly version as there might be different byte alignments / data type sizes.
2. the emulation core is a big state machine, when being refactored or enhanced the amount of state variables changes and so is the snapshot format. To make it short snapshot format might break already with a next version of vAmiga ... 

how to circumvent these two obstacles ...

+ you have to take the snapshots with web version (not the mac version) ... export them in the snapshot browser (there is a download icon in the right bottom corner of each snapshot image) and feed that file instead of the adf files to the player ...

+ we could provide a fixed version of vAmigaWeb which is never updated to a location lets say https://vamigaweb.github.io/v22 or alternatively you might fork the https://vamigaweb.github.io repo and prevent sync/pulls

**Comment by rhinoid** [comment]
I see your point that it might be tricky.
But wouldn't most of your points be alleviated with an export format which could even be an 'open spec' which all emulators abide by?  "The Open Amiga Snapshot Spec".
So some kind of json file (or binary if json is too big) with the processor registetrs, the flags, the programcounter, the chips, the memory, etc etc...  
This way the format is separated from the way it is stored in the emulator. It's more like an Amiga snapshot than an emulator snapshot so I'd think the spec could actually be completely supported by all emulators.

But agreed it does seem like a bit more work than I initially thought. 

It would open the door for a lot of 'linking to specific game/demo memories' from all over the internet.
Now the only thing you can do it: remember game X... and then they need to set controls, figure the game out, get good at the game, and then at some point get to that spot. Snapshots remove all that burden and offer more nostalgic opportunities. Even for people who just contributed to a small part:   "Look this is the work I did on the endboss of level 18..." you click and you're there. ;^D

The web version of vAmiga already gets closer to this point of making access to memories easier. No need to install software if it's just for that quick buzz.

**Comment by mithrendal** [comment]
the 'open spec' sounds interesting ... we have to ask @dirkwhoffmann what he thinks about it 

what about a never updating v22 version ... in the meantime ? Should we give it a go ? I can copy the current version to https://vamigaweb.github.io/v22 and leave it there forever untouched ... so that the snapshot format will never change when using that version

**Comment by dirkwhoffmann** [comment]
> the 'open spec' sounds interesting ... 

Sure, it sounds like a straightforward idea, and I tried something like this for VirtualC64 about 10 years ago. Bottom line: it didn't work reliably. Knowing all the registers, Ram, etc. is not enough to derive a consistent emulator state. At best you would end up with a "works most of the time" solution, which is not what I am aiming for.

**Comment by mithrendal** [comment]
ok I forked the current regular version from https://vamigaweb.github.io to https://vamigaweb.github.io/v22

and then I did the following changes to v22:

* changed the /v22/js/vAmigaWeb_player.js to point to /v22/
* changed the name of the PWA manifest to /v22/

set the branch protection rule for the v22 repo to
![image](https://user-images.githubusercontent.com/17108995/204109048-35b6ce36-c8bc-4b47-b939-b2067408f343.png)

theoretically this version could now last forever with a stable snapshot format which never breaks due to a change ... 😎

... although in reality we must face that this stable version could also break somewhen in 5 to 10 years from now when browser apis introduce breaking changes ... in that case someone possibly me has to adapt the javascript part which might be more or less easy ... in case the wasm ABI introduces breaking changes maybe in 10 to 20 years then fixing will be more difficult, maybe impossible   ( but if that happens a lot of other by then historic websites will have trouble too. Maybe they will never let it break ? ok when this happens someone has to play all the games again and make new snapshots 😂 )

**Comment by rhinoid** [comment]
Oh.... I hope I'm not causing too much pain and problems. It was really meant as a suggestion / attempt at a good idea.
I thought it could really make the embedded player take off because it lowers the bar so much for people casually passing by the webpage. 

But it shouldn't cause a maintenance nightmare. I've been too long a coder to know that you really need to do everything to keep stuff simple, because complexity is inevitable. So adding complexity on purpose is really a big deal.

**Comment by nicolasbauw** [comment]
> At the moment it looks like snapshots are ephemeral, and they go away the moment the machine is turned off. So the use case right now is a way to skip back in a game and have infinite tries.
> 
> The use case I'm thinking of is to have snapshots of the various interesting points in my game(s) so people can jump in without first playing everything. 

I use this feature a lot in vAmigaWeb. Game ready to play instantly. I also use that in place of the in-game saves.
I'd love to have the same feature in vAmiga : keeping the snapshots.

**Comment by mithrendal** [comment]
@nicolasbauw I have found a way how to do it in vAmiga.app for MacOS.  You can keep a snapshot with vAmiga.app when saving it to the mac filesystem.

you choose from menu 

`Machine`->`Save` 

this way you can drag multiple saved snapshots (one by one) back into a cold started vAmiga.app. They will then also be shown in the snapshot browser. 

Maybe this helps already a bit?
